intermediate_source/rpc_tutorial.rst

```diff
@@ -0,0 +1,529 @@
+optimizer takes a list of ``RRefs``, one ``RRef`` per model parameter for both
+local and remote parameters. The helper function is pretty simple, just call
+``Module.parameters()`` and creates a local ``RRef`` on each of the parameters.
+
+.. code:: python
+
+    def _parameter_rrefs(module):
+      param_rrefs = []
+      for param in module.parameters():
+          param_rrefs.append(RRef(param))
+      return param_rrefs
+
+Then, as the ``RNNModel`` contains three sub-modules, we need to call
+``_parameter_rrefs`` three times, and wrap that into another helper function.
+
+.. code:: python
+
+    class RNNModel(nn.Module):
+        ...
+        def parameter_rrefs(self):
+            remote_params = []
+            # get RRefs of embedding table
+            remote_params.extend(_remote_method(_parameter_rrefs, self.emb_table_rref))
+            # create RRefs for local parameters
+            remote_params.extend(_parameter_rrefs(self.rnn))
+            # get RRefs of decoder
+            remote_params.extend(_remote_method(_parameter_rrefs, self.decoder_rref))
+            return remote_params
+
+Now, we are ready to implement the training loop. After initializing the model
+arguments, we create the ``RNNModel`` and the ``DistributedOptimizer``. The
+distributed optimizer will take a list of parameter ``RRefs``, find all distinct
+owner workers, and create the given local optimizer (i.e., ``SGD`` in this case)
+on each of the owner worker using the given arguments (i.e., ``lr=0.05``).
+
+In the training loop, it first creates a distributed autograd context, which
+will help the distributed autograd engine to find gradients and involved RPC
+send/recv functions. Then, it kicks off the forward pass as if it is a local
+model, and run the distributed backward pass. For the distributed backward, you
+only need to specify a list of roots, in this case, it is the loss ``Tensor``.
+The distributed autograd engine will traverse the distributed graph
+automatically and write gradients properly. Next, it runs the ``step``
+API on the distributed optimizer, which will reach out to all involved local
+optimizers to update model parameters. Compared to local training, one minor
+difference is that you don't need to run ``zero_grad()`` because each autograd
+context has dedicated space to store gradients, and as we create a context
+per iteration, those gradients from different iterations will not accumulate to
+the same set of ``Tensors``.
+
+.. code:: python
+
+    def run_trainer():
+        batch = 5
+        ntoken = 10
+        ninp = 2
+
+        nhid = 3
+        nindices = 3
+        nlayers = 4
+        hidden = (
+            torch.randn(nlayers, nindices, nhid),
+            torch.randn(nlayers, nindices, nhid)
+        )
+
+        model = rnn.RNNModel('ps', ntoken, ninp, nhid, nlayers)
+
+        # setup distributed optimizer
+        opt = DistributedOptimizer(
+            optim.SGD,
```

Comment:
we have support for other optimizers such as `Adam` right?

Reply:
Yes, any local optimizers should work I think. Let me mention that in the docstring